# Import dan preparasi data

In [29]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [30]:
pip install luwiji

In [31]:
import numpy as np
import pandas as pd

from luwiji.text_proc import illustration
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer, TfidfVectorizer

In [32]:
data = pd.read_csv('data.csv')
data.head()

Unnamed: 0                  userName  \
0           0               Gadri Bandu   
1           1  070_Putri Azzahra Nurdin   
2           2       Novrealita Setiyana   
3           3                auah gelap   
4           4               Adi Nugroho   

                                             content  score  
0  Saya kasih bintang tiga dulu soalnga gemenya b...      3  
1  Bagus bgtt. Dari segi story, grafik sampe puzz...      5  
2  Sejauh ini aku main aku suka banget sama model...      5  
3  Semuanya sudah sempurna tapi Sangat disayangka...      3  
4  Grafik bagus,story bagus,terutama archon quest...      5

In [33]:
tex = data.drop(['Unnamed: 0', 'userName', 'score'], axis=1)
tex

content
0    Saya kasih bintang tiga dulu soalnga gemenya b...
1    Bagus bgtt. Dari segi story, grafik sampe puzz...
2    Sejauh ini aku main aku suka banget sama model...
3    Semuanya sudah sempurna tapi Sangat disayangka...
4    Grafik bagus,story bagus,terutama archon quest...
..                                                 ...
995  Masih bingung sih dengan cara game ini ngerend...
996  game nya seru dari segi story. saran buat Hoyo...
997  Game ini sangat bagus dengan map yg sangat lua...
998  Sangat seru dan asik untuk dimainkan grafik hd...
999  Memuat game terlalu lama . Mau masuk game aja ...

[1000 rows x 1 columns]

In [37]:
tex['content'] = tex['content'].str.lower()
teks_use = tex.sample(n=5) #mengambil 5 data random pada dataframe content untuk dianalisis TF-IDF
teks_use

content
252  game nya udah bagus overall, dari grafik sudah...
750  game nya seru banget tapi kadang² aku gk bisa....
130  saya udah muak ya ama jaringan selalu aja mera...
20   bagus dari segi grafik nya.. saran buat player...
724  main dari 2020, tpi ngeliat update kali ini ya...

In [38]:
array_teks = np.array(teks_use)
array_teks

array([['game nya udah bagus overall, dari grafik sudah memadai, backsound juga sudah bagus disaat momen yg tepat, tapi masalah nya koq lama kelamaan ini game bikin ngelag dan nguras baterai yg sangat banyak. baru di mainin 10 menit, grafik sudah low walaupun hp ku ini sudah cukup melewati rata rata spek yg disarankan, hp ku baru main 10 menit dari baterai nya 100% jadi 78%, terus bikin panas, dan panas itu membuat hp ku ini semakin ngelag... untuk developer tolong gamenya di kompres lebih ringan mb nya'],
       ['game nya seru banget tapi kadang² aku gk bisa.juga pas login lama banget'],
       ['saya udah muak ya ama jaringan selalu aja merah padahal jelas jelas aslinya itu bagus juga banyak bug. bisa gak jaringannya itu dibenerin? gue algi ngedomain jadi mati kalau udah dibenerin nanti saya ganti ulasanya'],
       ['bagus dari segi grafik nya.. saran buat player yg mau donlaud: kalau mau donlaud game ini mending liat memori sm ram dulu deh, gede buanget masalahnya, di pc aja bisa 

# Melakukan split pada data teks yang telah terpilih

In [70]:
import nltk
from nltk.tokenize import sent_tokenize
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [60]:
# Contoh data teks dalam array NumPy dalam format bytes
array_teks = np.array(['game nya udah bagus overall, dari grafik sudah memadai, backsound juga sudah bagus disaat momen yg tepat, tapi masalah nya koq lama kelamaan ini game bikin ngelag dan nguras baterai yg sangat banyak. baru di mainin 10 menit, grafik sudah low walaupun hp ku ini sudah cukup melewati rata rata spek yg disarankan, hp ku baru main 10 menit dari baterai nya 100% jadi 78%, terus bikin panas, dan panas itu membuat hp ku ini semakin ngelag... untuk developer tolong gamenya di kompres lebih ringan mb nya',
 'game nya seru banget tapi kadang² aku gk bisa.juga pas login lama banget',
 'saya udah muak ya ama jaringan selalu aja merah padahal jelas jelas aslinya itu bagus juga banyak bug. bisa gak jaringannya itu dibenerin? gue algi ngedomain jadi mati kalau udah dibenerin nanti saya ganti ulasanya',
 'bagus dari segi grafik nya.. saran buat player yg mau donlaud: kalau mau donlaud game ini mending liat memori sm ram dulu deh, gede buanget masalahnya, di pc aja bisa 72gb, di android bisa 40gb (buat player lama/quest kelar) udah sih, segini aja, dan boleh gak kasih quest monstadt lagi..? kangennn venti',
 'main dari 2020, tpi ngeliat update kali ini yang makin bengkak, kayanya emang udah harus pensi'])

# Menerapkan sent_tokenize pada setiap elemen dalam array
teks_split = [sent_tokenize(teks) for teks in array_teks]

# Hasilnya adalah array yang berisi kalimat-kalimat terpisah
teks_split

[['game nya udah bagus overall, dari grafik sudah memadai, backsound juga sudah bagus disaat momen yg tepat, tapi masalah nya koq lama kelamaan ini game bikin ngelag dan nguras baterai yg sangat banyak.',
  'baru di mainin 10 menit, grafik sudah low walaupun hp ku ini sudah cukup melewati rata rata spek yg disarankan, hp ku baru main 10 menit dari baterai nya 100% jadi 78%, terus bikin panas, dan panas itu membuat hp ku ini semakin ngelag... untuk developer tolong gamenya di kompres lebih ringan mb nya'],
 ['game nya seru banget tapi kadang² aku gk bisa.juga pas login lama banget'],
 ['saya udah muak ya ama jaringan selalu aja merah padahal jelas jelas aslinya itu bagus juga banyak bug.',
  'bisa gak jaringannya itu dibenerin?',
  'gue algi ngedomain jadi mati kalau udah dibenerin nanti saya ganti ulasanya'],
 ['bagus dari segi grafik nya.. saran buat player yg mau donlaud: kalau mau donlaud game ini mending liat memori sm ram dulu deh, gede buanget masalahnya, di pc aja bisa 72gb, di 

# Proses TF-IDF

In [62]:
teks_gabung = [' '.join(kalimat) for kalimat in teks_split]

In [63]:
tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(teks_gabung)

In [65]:
tfidf_matrix

<5x132 sparse matrix of type '<class 'numpy.float64'>'
	with 163 stored elements in Compressed Sparse Row format>

In [66]:
print(tfidf_matrix)

  (0, 78)	0.08712690621877124
  (0, 105)	0.08712690621877124
  (0, 67)	0.08712690621877124
  (0, 62)	0.08712690621877124
  (0, 39)	0.08712690621877124
  (0, 121)	0.08712690621877124
  (0, 29)	0.08712690621877124
  (0, 125)	0.08712690621877124
  (0, 112)	0.08712690621877124
  (0, 81)	0.08712690621877124
  (0, 48)	0.07029343934169299
  (0, 97)	0.17425381243754248
  (0, 120)	0.08712690621877124
  (0, 5)	0.08712690621877124
  (0, 49)	0.07029343934169299
  (0, 1)	0.08712690621877124
  (0, 71)	0.07029343934169299
  (0, 33)	0.08712690621877124
  (0, 116)	0.08712690621877124
  (0, 104)	0.17425381243754248
  (0, 79)	0.08712690621877124
  (0, 25)	0.08712690621877124
  (0, 64)	0.2613807186563137
  (0, 46)	0.2613807186563137
  (0, 128)	0.08712690621877124
  :	:
  (3, 47)	0.08773765976624896
  (3, 66)	0.08773765976624896
  (3, 131)	0.10569652025165721
  (3, 43)	0.10569652025165721
  (3, 27)	0.08773765976624896
  (3, 13)	0.08773765976624896
  (3, 123)	0.07380768536041553
  (3, 94)	0.0877376597662489

In [73]:
vocab = tfidf.get_feature_names_out()
vocab

array(['10', '100', '2020', '40gb', '72gb', '78', 'aja', 'aku', 'algi',
       'ama', 'android', 'aslinya', 'backsound', 'bagus', 'banget',
       'banyak', 'baru', 'baterai', 'bengkak', 'bikin', 'bisa', 'boleh',
       'buanget', 'buat', 'bug', 'cukup', 'dan', 'dari', 'deh',
       'developer', 'di', 'dibenerin', 'disaat', 'disarankan', 'donlaud',
       'dulu', 'emang', 'gak', 'game', 'gamenya', 'ganti', 'gede', 'gk',
       'grafik', 'gue', 'harus', 'hp', 'ini', 'itu', 'jadi', 'jaringan',
       'jaringannya', 'jelas', 'juga', 'kadang²', 'kalau', 'kali',
       'kangennn', 'kasih', 'kayanya', 'kelamaan', 'kelar', 'kompres',
       'koq', 'ku', 'lagi', 'lama', 'lebih', 'liat', 'login', 'low',
       'main', 'mainin', 'makin', 'masalah', 'masalahnya', 'mati', 'mau',
       'mb', 'melewati', 'memadai', 'membuat', 'memori', 'mending',
       'menit', 'merah', 'momen', 'monstadt', 'muak', 'nanti',
       'ngedomain', 'ngelag', 'ngeliat', 'nguras', 'nya', 'overall',
       'padahal', 'pan

In [74]:
pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf.get_feature_names_out(), index=teks_gabung)

10       100  \
game nya udah bagus overall, dari grafik sudah ...  0.174254  0.087127   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.000000  0.000000   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.000000   
main dari 2020, tpi ngeliat update kali ini yan...  0.000000  0.000000   

                                                        2020      40gb  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.000000   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.000000  0.000000   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.131008   
main dari 2020, tpi ngeliat update kali ini yan...  0.268556  0.000000   

                                                        72gb        78  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.087127   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.000000  0.000000   
bagus dari segi grafik nya.. saran buat player ...  0.131008  0.000000   
main dari 2020, tpi ngeliat update kali ini yan...  0.000000  0.000000   

                                                         aja       aku  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.000000   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.278493   
saya udah muak ya ama jaringan selalu aja merah...  0.131796  0.000000   
bagus dari segi grafik nya.. saran buat player ...  0.211393  0.000000   
main dari 2020, tpi ngeliat update kali ini yan...  0.000000  0.000000   

                                                        algi       ama  ...  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.000000  ...   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000  ...   
saya udah muak ya ama jaringan selalu aja merah...  0.163357  0.163357  ...   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.000000  ...   
main dari 2020, tpi ngeliat update kali ini yan...  0.000000  0.000000  ...   

                                                         tpi      udah  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.049086   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.000000  0.184065   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.073808   
main dari 2020, tpi ngeliat update kali ini yan...  0.268556  0.151300   

                                                    ulasanya     untuk  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.087127   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.163357  0.000000   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.000000   
main dari 2020, tpi ngeliat update kali ini yan...  0.000000  0.000000   

                                                      update     venti  \
game nya udah bagus overall, dari grafik sudah ...  0.000000  0.000000   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.000000  0.000000   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.131008   
main dari 2020, tpi ngeliat update kali ini yan...  0.268556  0.000000   

                                                    walaupun        ya  \
game nya udah bagus overall, dari grafik sudah ...  0.087127  0.000000   
game nya seru banget tapi kadang² aku gk bisa.j...  0.000000  0.000000   
saya udah muak ya ama jaringan selalu aja merah...  0.000000  0.163357   
bagus dari segi grafik nya.. saran buat player ...  0.000000  0.000000   
main dari 2020, tpi ngeliat update kali ini yan...  0.000000  0.000000   

                       

In [83]:
dokumen_ke = 1 # Indeks data yang ingin diperiksa

indeks_kata_tertinggi = tfidf_matrix[dokumen_ke].argmax()

kata_tertinggi = vocab[indeks_kata_tertinggi]

bobot_tertinggi = tfidf_matrix[dokumen_ke, indeks_kata_tertinggi]

print(f"Kata dengan bobot tertinggi dalam dokumen {dokumen_ke} adalah '{kata_tertinggi}' dengan bobot {bobot_tertinggi:.2f}")

Kata dengan bobot tertinggi dalam dokumen 1 adalah 'banget' dengan bobot 0.56


**Terdapat 132 kolom yang mewakili kata dari 5 data yang terpilih, dan dari kelima data tersebut, kata dengan bobot paling tinggi terdapat pada kata "banget" di data indeks ke-1 atau data kedua dengan nilai bobot sebesar 0.56.**

Kata dengan bobot tertinggi dalam dokumen 0 adalah 'sudah' dengan bobot 0.35

Kata dengan bobot tertinggi dalam dokumen 1 adalah 'banget' dengan bobot 0.56

Kata dengan bobot tertinggi dalam dokumen 2 adalah 'dibenerin' dengan bobot 0.33

Kata dengan bobot tertinggi dalam dokumen 3 adalah 'buat' dengan bobot 0.26

Kata dengan bobot tertinggi dalam dokumen 4 adalah '2020' dengan bobot 0.27